File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2020/Y20D03.kt
package org.gristle.pdxpuzzles.advent.y2020

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.objects.Coord
import org.gristle.pdxpuzzles.utilities.objects.toGrid

class Y20D03(input: String) : Day {

    val grid = input.toGrid { it == '#' }

    fun solve(right: Int = 3, down: Int = 1) =
        (down until grid.height step down).foldIndexed(0) { i, acc, y ->
            acc + if (grid[Coord(((i + 1) * right) % grid.width, y)]) 1 else 0
        }

    override fun part1() = solve()

    override fun part2() = listOf(
        solve(1),
        solve(3),
        solve(5),
        solve(7),
        solve(1, 2)
    ).fold(1L, Long::times)
}

fun main() = Day.runDay(Y20D03::class)

//    Class creation: 24ms
//    Part 1: 294 (1ms)
//    Part 2: 5774564250 (0ms)
//    Total time: 26ms

File: rust/advent/src/bin/2020/y20d03.rs
use advent::utilities::get_input::get_input;
use utilities::structs::coord::Coord2U;
use utilities::structs::grid::Grid2;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input = Grid2<bool>;
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(20, 3).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    let slope = Grid2::new2d_map_str(input, |c| c == '#').unwrap();
    slope
}

fn solve(slope: &Input, right: usize, down: usize) -> Output {
    (down..slope.height()).step_by(down).enumerate()
        .fold(0, |acc, (idx, y)| {
            acc + if slope[Coord2U::new2d((idx + 1) * right % slope.width(), y)] { 1 } else { 0 }
        })
}

fn part1(input: &Input) -> Output {
    solve(input, 3, 1)
}

fn part2(input: &Input) -> Output {
    [
        solve(input, 1, 1),
        solve(input, 3, 1),
        solve(input, 5, 1),
        solve(input, 7, 1),
        solve(input, 1, 2),
    ]
        .into_iter()
        .fold(1, |acc, x| acc * x)
}

#[test]
fn default() {
    let input = get_input(20, 3).unwrap();
    let input = parse_input(&input);
    assert_eq!(294, part1(&input));
    assert_eq!(5774564250, part2(&input));
}

// Input parsed (61μs)
// 1. 294 (23μs)
// 2. 5774564250 (78μs)
// Total: 164μs